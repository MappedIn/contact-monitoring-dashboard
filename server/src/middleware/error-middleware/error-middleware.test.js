jest.mock('../../logger');

const logger = require('../../logger');
const { celebrate, Joi } = require('celebrate');
const express = require('express');
const supertest = require('supertest');

const ClientError = require('./client-error');
const errorMiddleware = require('./error-middleware');

describe('error middleware', () => {
	describe('basic cases without express', () => {
		test('calls next when there error is falsey', () => {
			const next = jest.fn();
			errorMiddleware(null, null, null, next);
			expect(next).toHaveBeenCalledWith();
		});
	});

	describe('tests using express', () => {
		let app;
		const clientError = new ClientError({
			code: ClientError.Types.CONFLICT,
			message: 'client error',
		});

		beforeAll(() => {
			app = express();
			app.get(
				'/celebrate-error',
				celebrate({
					query: Joi.object().keys({
						key: Joi.string().required(),
					}),
				}),
				(req, res) => {
					res.sendStatus(204);
				},
			);

			app.get('/client-error', () => {
				throw clientError;
			});

			app.get('/unknown-error', () => {
				throw new Error();
			});

			app.use(errorMiddleware);
		});

		test('calls logger.error with the error', async () => {
			await supertest(app)
				.get('/client-error')
				.expect(ClientError.Types.CONFLICT);

			expect(logger.error).toHaveBeenCalledWith(clientError);
		});

		test('captures ClientError', async () => {
			const response = await supertest(app)
				.get('/client-error')
				.expect(ClientError.Types.CONFLICT);

			expect(response.text).toMatchSnapshot();
		});

		test('captures celebrate error', async () => {
			const response = await supertest(app)
				.get('/celebrate-error')
				.expect(400);

			expect(response.text).toMatchSnapshot();
		});

		test('captures unexpected error', async () => {
			const response = await supertest(app)
				.get('/unknown-error')
				.expect(500);

			expect(response.text).toMatchSnapshot();
		});
	});
});
