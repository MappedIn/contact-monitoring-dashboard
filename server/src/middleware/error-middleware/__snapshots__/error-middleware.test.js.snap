// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error middleware tests using express captures ClientError 1`] = `"{\\"error\\":{\\"code\\":409,\\"message\\":\\"client error\\"}}"`;

exports[`error middleware tests using express captures celebrate error 1`] = `"{\\"error\\":{\\"joi\\":{\\"isJoi\\":true,\\"name\\":\\"ValidationError\\",\\"details\\":[{\\"message\\":\\"\\\\\\"key\\\\\\" is required\\",\\"path\\":[\\"key\\"],\\"type\\":\\"any.required\\",\\"context\\":{\\"key\\":\\"key\\",\\"label\\":\\"key\\"}}],\\"_object\\":{}},\\"meta\\":{\\"source\\":\\"query\\"}}}"`;

exports[`error middleware tests using express captures unexpected error 1`] = `"{\\"error\\":{\\"code\\":500,\\"message\\":\\"A server error has occured\\"}}"`;
